File: jetbrains/src/main/kotlin/deplens/lang/dart/DartDepLensInlayProvider.kt
package deplens.lang.dart

import com.intellij.codeInsight.hints.declarative.InlayTreeSink
import com.intellij.openapi.editor.Document
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import deplens.lang.BaseDepLensInlayProvider
import deplens.utils.inlay.PubInlayUtils
import deplens.utils.resolver.DartImportResolver

class DartDepLensInlayProvider : BaseDepLensInlayProvider() {

    override val usesWholeFileCollection: Boolean = true

    override fun isFileSupported(file: PsiFile): Boolean = file.name.endsWith(".dart")

    override fun collectElement(file: PsiFile, element: PsiElement, sink: InlayTreeSink) = Unit

    override fun collectFile(file: PsiFile, sink: InlayTreeSink) {
        val document = file.viewProvider.document ?: return

        for (lineNumber in 0 until document.lineCount) {
            val lineText = document.getLineText(lineNumber)
            val dep = DartImportResolver.getDepName(lineText) ?: continue
            if (DartImportResolver.isLocalImport(dep) || !DartImportResolver.isPackageImport(dep)) {
                continue
            }

            val pkg = DartImportResolver.getPkgName(dep)
            val offset = document.getLineEndOffset(lineNumber)
            PubInlayUtils.addPubDepInlay(file, sink, pkg, offset)
        }
    }

    private fun Document.getLineText(lineNumber: Int): String {
        val startOffset = getLineStartOffset(lineNumber)
        val endOffset = getLineEndOffset(lineNumber)
        return getText(com.intellij.openapi.util.TextRange(startOffset, endOffset))
    }
}

File: vscode/src/lang/dart/DartDepLensInlayProvider.ts
import * as vscode from "vscode";
import { BaseDepLensInlayProvider } from "../BaseDepLensInlayProvider";
import { PubInlayUtils } from "../../utils/inlay/PubInlayUtils";
import { DartImportResolver } from "../../utils/resolver/DartImportResolver";
import { GithubRepoInfoService } from "../../utils/service/GithubRepoInfoService";
import { PubPkgInfoService } from "../../utils/service/PubPkgInfoService";

export class DartDepLensInlayProvider extends BaseDepLensInlayProvider {
  constructor() {
    super();
    PubPkgInfoService.getInstance().onDidUpdatePackageInfo(() => {
      this.emitter.fire();
    });
    GithubRepoInfoService.getInstance().onDidUpdateRepoInfo(() => {
      this.emitter.fire();
    });
  }

  protected isFileSupported(document: vscode.TextDocument): boolean {
    return document.languageId === "dart";
  }

  protected async provideInlayHintsForDocument(
    document: vscode.TextDocument,
    range: vscode.Range,
    token: vscode.CancellationToken,
  ): Promise<vscode.InlayHint[]> {
    const hints: vscode.InlayHint[] = [];

    for (let line = range.start.line; line <= range.end.line; line++) {
      if (token.isCancellationRequested) {
        return hints;
      }

      const text = document.lineAt(line).text;
      if (!DartImportResolver.isImport(text)) {
        continue;
      }

      const dep = DartImportResolver.getDepName(text);
      if (
        !dep ||
        DartImportResolver.isLocalImport(dep) ||
        !DartImportResolver.isPackageImport(dep)
      ) {
        continue;
      }

      const pkg = DartImportResolver.getPkgName(dep);
      const position = new vscode.Position(line, text.length);
      PubInlayUtils.addPubDepInlay(hints, pkg, position);
    }

    return hints;
  }
}
